### PR TITLE
Add rcutils_set_env function

### PR DESCRIPTION
**Pull request by cottsay**
This change adds a simple function for setting or un-setting an environment variable for the process.

The existing environment variable functions in rcutils are under a header called `get_env.h`. It might be good to deprecate that header and move the functionality to the more generically-named `env.h`. I'd love to hear feedback on that.

If we decide to keep a separate `get_env.h` header moving forward, I should probably rename this one to `set_env.h`.

* Linux [Build Status](http://ci.ros2.org/job/ci_linux/10767/)
* Linux-aarch64 [Build Status](http://ci.ros2.org/job/ci_linux-aarch64/6159/)
* macOS [Build Status](http://ci.ros2.org/job/ci_osx/8765/)
* Windows [Build Status](http://ci.ros2.org/job/ci_windows/10657/)

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -291,6 +292,16 @@ if(BUILD_TESTING)
     target_link_libraries(test_get_env ${PROJECT_NAME})
   endif()
 
+  rcutils_custom_add_gtest(test_env test/test_env.cpp
+    ENV
+      EMPTY_TEST=
```
Comment (ahcorde) [inline CMakeLists.txt:297]: why are you adding these env vars here? They are not used in the test

**Comment by clalancette** [comment]
> The existing environment variable functions in rcutils are under a header called `get_env.h`. It might be good to deprecate that header and move the functionality to the more generically-named `env.h`. I'd love to hear feedback on that.

Yes, I'd be all for that.

File: include/rcutils/env.h [hunk 2]
```diff
@@ -0,0 +1,47 @@
...
+/// Set or un-set a process-scoped environment variable.
+/**
+ * This function modifies the environment variables for the current process.
+ *
+ * \param[in] env_name Name of the environment variable to modify.
+ * \param[in] env_value Value to set the environment variable to, or NULL to un-set.
+ * \return `True` if success
+ * \return `False` if env_name is invalid or NULL
+ * \return `False` on failure
+ */
```
Comment (clalancette) [inline include/rcutils/env.h:37]: I think its important to point out the thread-safetyness of this function, both with respect to itself and with respect to concurrent changes when get_env when running.  i.e. the ongoing saga of https://github.com/ros2/rcutils/issues/242 .

File: src/env.c [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+  if (NULL == env_value) {
+    env_value = "";
+  }
+  if (0 != _putenv_s(env_name, env_value)) {
+    return false;
+  }
+#else
+  if (NULL == env_value) {
+    if (0 != unsetenv(env_name)) {
+      return false;
```
Comment (clalancette) [inline src/env.c:43]: Instead of just swallowing the error here, I think it would be a better idea to call strerror to get the actual error and then `RCUTILS_SET_ERROR_MSG...` to set it.  The same goes above (though the call on Windows is different) and below.

File: src/env.c [hunk 4]
```diff
@@ -0,0 +1,57 @@
...
+rcutils_set_env(const char * env_name, const char * env_value)
+{
+  RCUTILS_CHECK_FOR_NULL_WITH_MSG(
+    env_name, "env_name is null", return false);
+
+#ifdef _WIN32
+  if (NULL == env_value) {
+    env_value = "";
+  }
+  if (0 != _putenv_s(env_name, env_value)) {
```
Comment (clalancette) [inline src/env.c:37]: For Unix, we are careful to use `setenv` as it makes copies of the strings.  From https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/putenv-s-wputenv-s?view=vs-2019 , it isn't clear to me whether Windows does the same.  Do you know whether it makes copies or not?

File: CMakeLists.txt (hunk 1 above)
Comment (cottsay) [reply CMakeLists.txt]: I was originally going to merge the get_env stuff in this PR - I no longer think the timing is right. Removed in a11900b.

**Comment by cottsay** [comment]
> > The existing environment variable functions in rcutils are under a header called `get_env.h`. It might be good to deprecate that header and move the functionality to the more generically-named `env.h`. I'd love to hear feedback on that.
> 
> Yes, I'd be all for that.

It's too late to deprecate get_env.h in Foxy, but we can at least make the new header compatible in Foxy. We can deprecate `get_env.h` in G and remove it in H.

I added `get_env.h` to `env.h` in 9c67776.

**Comment by dirk-thomas** [comment]
What is the reason to add this function at all? What is it needed for?

**Comment by cottsay** [comment]
> What is the reason to add this function at all? What is it needed for?

I need to modify the environment for some feature testing in another package, where changing the environment for the entire test isn't desirable. This change clearly demonstrates that there isn't a foolproof cross-platform solution for doing it. If I'm going to do the work to create that mechanism and the accompanying tests, I thought it best to put it somewhere others can use it, especially because there are already some environment methods to accompany it.

**Comment by cottsay** [comment]
I had another thought while reading the MSDN docs: Since you can't create an empty environment variable on Windows (setting it to `""` just removes the variable), should I replicate that behavior on *nix as well, so that the behavior is always consistent across platforms?

**Comment by dirk-thomas** [comment]
I don't think prevent users on Linux to do what is possible on that platform is a good idea. I would rather document the difference in the docblock.

**Comment by cottsay** [comment]
> I don't think prevent users on Linux to do what is possible on that platform is a good idea. I would rather document the difference in the docblock.

Perfectly acceptable to me. Thanks for the feedback.

File: include/rcutils/env.h (hunk 2 above)
Comment (cottsay) [reply include/rcutils/env.h:40]: This is really great point. Both MSDN and setenv(3) call out the lack of thread safety. Neither differentiate between other instances of the same call and calls to a reader function, however. We might be able to get a firm answer for Linux by looking at the source, but unless someone finds something documented for Windows, it's probably safest to guard against current set operations.

From what I'm reading in #242, a resolution would mean guards in both get and set functions. If it's alright with you, I'd like to document the lack of thread safety in this PR, and let #242 address the thread safety issues between get and set operations. From the sounds of it, adding this function is the first step in that process anyway, and I believe that the signature of the function I'm adding here is adequate for implementing the new lock (unlike with `rcutils_get_env`).

Comment (clalancette) [reply include/rcutils/env.h:40]: > If it's alright with you, I'd like to document the lack of thread safety in this PR, and let #242 address the thread safety issues between get and set operations.

Yeah, that's exactly what I'm suggesting; just make sure to document it here.  We'll address the larger issues as part of #242

File: src/env.c (hunk 4 above)
Comment (cottsay) [reply src/env.c:37]: I assume that `setenv` is calling `malloc` then. Even `putenv`, which doesn't duplicate the string, can return `ENOMEM`, so I would assume that it is (re-)allocating memory as well. Is that a problem, or would documentation be sufficient? Are you aware of another way to do this where we can use an allocator?

I'll look into answering the question for Windows.

Comment (cottsay) [reply src/env.c:37]: I found some MSDN docs discussing the way environment variables are implemented in Windows: https://docs.microsoft.com/en-us/cpp/c-runtime-library/environ-wenviron?view=vs-2019

It is similar to Linux in that the key and value are stored in the same string, so calling a function that takes them separately can't re-use the strings and must be allocating new memory to store them combined.

I also tested whether the `_putenv`, which takes a single string like the Linux `putenv` function does, is allocating new memory, and it seems to be doing so and copying the string passed to `_putenv`. That said, I can't find any documentation explicitly stating that it happens like that all of the the time.

tl;dr - For our purposes, `_putenv_s` should act similar to `setenv` as far as ownership of the strings passed into it.

Comment (clalancette) [reply src/env.c:37]: > I assume that `setenv` is calling `malloc` then.

Yeah, I would guess so.  The man page for it says:

```
This function makes copies of the strings pointed to by name and value (by contrast with putenv(3)).
```

> Even `putenv`, which doesn't duplicate the string, can return `ENOMEM`, so I would assume that it is (re-)allocating memory as well.

My reading of glibc here is that `putenv` can indeed return ENOMEM.  In particular, this can happen if we are putting a *new* value into the environment, which can cause the environment to grow, which requires a realloc.  A replacement of an existing variable doesn't do any allocation at all for `putenv`.  For a new environment variable, `setenv` does two allocations: one to expand the environment and one to copy the string given by the user into glibc-controlled memory.  For a replacement, `setenv` does a single allocation to copy the string given by the user into glibc-controlled memory.

In any case, I don't think there is much we can do about that allocation; it happens deep in glibc.  What I'm aiming for here is consistency between platforms, which is where the question about Windows comes from.  Once we know the answer to that, we'll have a better idea on how to proceed.

Comment (clalancette) [reply src/env.c:37]: > tl;dr - For our purposes, `_putenv_s` should act similar to `setenv` as far as ownership of the strings passed into it.

Yeah, I just did a quick test here on Windows, and it does indeed seem like `_putenv_s` makes a copy of the environment string (at least, modifying it in place didn't change the returned value from getenv).  So I think we are good to go with this; just documentation that this function always *copies* into the environment is good enough for now.

File: src/env.c (hunk 3 above)
Comment (cottsay) [reply src/env.c:45]: Do you really think it's worth creating a buffer to receive the error message? The docs for all of these functions say that the only reasons they fail are if an invalid key was specified or there was a memory allocation error.

I added a message with the error code. If you'd really rather see the strerror, I can do that instead.

File: include/rcutils/env.h (hunk 2 above)
Comment (cottsay) [reply include/rcutils/env.h]: Done in 4ba9470

**Comment by cottsay** [comment]
> I would rather document the difference in the docblock.

Done in 4ba9470

File: src/env.c (hunk 4 above)
Comment (cottsay) [reply src/env.c:38]: Done in b697ce0

File: test/test_env.cpp [hunk 5]
```diff
@@ -0,0 +1,53 @@
...
+#include <string>
+
+#include "rcutils/env.h"
+#include "rcutils/get_env.h"
+
+TEST(TestEnv, test_set_env) {
+  const char * res = nullptr;
+
+  // Invalid cases
+  EXPECT_FALSE(rcutils_set_env(nullptr, nullptr));
```
Comment (ahcorde) [inline test/test_env.cpp:26]: You may want to call  `rcutils_reset_error();` when a error happens  otherwise you will show some overwritten errors in the terminal.

Comment (cottsay) [reply test/test_env.cpp:27]: Done in 8a6ee18

File: src/env.c [hunk 6]
```diff
@@ -0,0 +1,61 @@
...
+    return false;
+  }
+#else
+  if (NULL == env_value) {
+    if (0 != unsetenv(env_name)) {
+      RCUTILS_SET_ERROR_MSG_WITH_FORMAT_STRING("unsetenv failed: %d", errno);
+      return false;
+    }
+  } else {
+    if (0 != setenv(env_name, env_value, 1)) {
```
Comment (fujitatomoya) [inline src/env.c:49]: How about providing overwrite flag to user too?

**Comment by cottsay** [comment]
> Are you intending to land this for Foxy...?

It's a well-reviewed and tested addition to the API, and I'd like to use it in some tests in a downstream package, so it would be nice to land in Foxy. This package hasn't bumped to 1.0.0 yet, so I think there is another release that needs to happen anyway. Sound okay, @jacobperron?

File: src/env.c (hunk 6 above)
Comment (cottsay) [reply src/env.c:49]: There is no equivalent in the WIN32 API and I don't see a good way to do it without introducing a race, so I'd prefer to keep the API as uniform across platforms as possible.

**Comment by jacobperron** [comment]
@cottsay I think since this an addition (not modifying existing API), it's fine to land it for Foxy.

**Comment by cottsay** [comment]
> ...it's fine to land it for Foxy.

Grazie!
